iree/compiler/Dialect/VM/Conversion/UtilToVM/ConvertAlignmentOps.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "iree/compiler/Dialect/VM/Conversion/TypeConverter.h"
+#include "iree/compiler/Dialect/VM/Conversion/UtilToVM/ConvertUtilToVM.h"
+#include "iree/compiler/Dialect/VM/IR/VMOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+class UtilAlignOpConversion : public OpConversionPattern<IREE::Util::AlignOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      IREE::Util::AlignOp srcOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    IREE::Util::AlignOpAdaptor srcOperands(operands);
+    auto valueType = srcOperands.value().getType();
+
+    // (value + (alignment - 1)) & ~(alignment - 1)
+    if (valueType.isInteger(32)) {
```

Comment:
You can reduce the duplication here (same path for i32 and i64) by creating a helper function templated on the different VM ops, or by lowering to ops in higher level dialects (std/arith/etc., which aren't already specialized to a specific bit depth) and then relying on those dialects' lowering to the VM dialect.

For the helper function template approach, see this example: https://github.com/google/iree/blob/94d03f5ca9c942688471186d984448476e376436/iree/compiler/Dialect/VM/IR/VMOpFolders.cpp#L181-L218

Reply:
Helper function yes. not sure about the need for template (or what to template on...)

Reply:
```c++
if (valueType.isInteger(32))
  insertAlignOps<IREE::VM::ConstI32Op, IREE::VM::SubI32Op, IREE::VM::AddI32Op, ...>(args...);
} else if (valueType.isInteger(64)) {
  insertAlignOps<IREE::VM::ConstI64Op, IREE::VM::SubI64Op, IREE::VM::AddI64Op, ...>(args...);
}
```